<a href="https://colab.research.google.com/github/Amuktha28/ML-PROJECT/blob/main/Myproject.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import StratifiedKFold, train_test_split
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn.compose import ColumnTransformer
from sklearn.metrics import accuracy_score


In [ ]:

test = pd.read_csv('/content/test (2).csv')
train = pd.read_csv('/content/train (2).csv')

In [ ]:
train.isnull().sum()

id                                            0
Name                                          0
Gender                                        0
Age                                           0
City                                          0
Working Professional or Student               0
Profession                                36630
Academic Pressure                        112803
Work Pressure                             27918
CGPA                                     112802
Study Satisfaction                       112803
Job Satisfaction                          27910
Sleep Duration                                0
Dietary Habits                                4
Degree                                        2
Have you ever had suicidal thoughts ?         0
Work/Study Hours                              0
Financial Stress                              4
Family History of Mental Illness              0
Depression                                    0
dtype: int64

In [ ]:
null_cols = train.isnull().sum()
num_null_cols = len(null_cols[null_cols > 0])
print(f"Number of columns with null values: {num_null_cols}")

Number of columns with null values: 9


In [ ]:
# prompt: convert null columns if it is object use mode imputer if it is int8,flota32 use mean imputer

import pandas as pd
import numpy as np
from sklearn.impute import SimpleImputer

# Assuming 'train' DataFrame is already loaded as shown in the previous code

# Identify columns with null values
null_cols = train.columns[train.isnull().any()]

# Iterate through columns with null values
for col in null_cols:
    if train[col].dtype == 'object':
        # Use mode imputation for object type columns
        imputer = SimpleImputer(strategy='most_frequent')
        # Reshape the output of fit_transform to 1D
        train[col] = imputer.fit_transform(train[[col]]).ravel()
    elif train[col].dtype in ['int8', 'float32', 'float64', 'int64', 'int32']:
        # Use mean imputation for numerical type columns
        imputer = SimpleImputer(strategy='mean')
        # Reshape the output of fit_transform to 1D
        train[col] = imputer.fit_transform(train[[col]]).ravel()
    else:
        print(f"Column '{col}' has unsupported data type: {train[col].dtype}")

In [ ]:
train.isnull().sum()

id                                       0
Name                                     0
Gender                                   0
Age                                      0
City                                     0
Working Professional or Student          0
Profession                               0
Academic Pressure                        0
Work Pressure                            0
CGPA                                     0
Study Satisfaction                       0
Job Satisfaction                         0
Sleep Duration                           0
Dietary Habits                           0
Degree                                   0
Have you ever had suicidal thoughts ?    0
Work/Study Hours                         0
Financial Stress                         0
Family History of Mental Illness         0
Depression                               0
dtype: int64

In [ ]:
null_cols = train.isnull().sum()
num_null_cols = len(null_cols[null_cols > 0])
print(f"Number of columns with null values: {num_null_cols}")

Number of columns with null values: 0


In [ ]:
train.dtypes

id                                         int64
Name                                      object
Gender                                    object
Age                                      float64
City                                      object
Working Professional or Student           object
Profession                                object
Academic Pressure                        float64
Work Pressure                            float64
CGPA                                     float64
Study Satisfaction                       float64
Job Satisfaction                         float64
Sleep Duration                            object
Dietary Habits                            object
Degree                                    object
Have you ever had suicidal thoughts ?     object
Work/Study Hours                         float64
Financial Stress                         float64
Family History of Mental Illness          object
Depression                                 int64
dtype: object

In [ ]:
# prompt: check all columns in train id dtype is object convert them into int using encoding

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer

test = pd.read_csv('/content/test (2).csv')
train = pd.read_csv('/content/train (2).csv')

# Identify columns with null values
null_cols = train.columns[train.isnull().any()]

# Iterate through columns with null values
for col in null_cols:
    if train[col].dtype == 'object':
        # Use mode imputation for object type columns
        imputer = SimpleImputer(strategy='most_frequent')
        train[col] = imputer.fit_transform(train[[col]]).ravel()
    elif train[col].dtype in ['int8', 'float32', 'float64', 'int64', 'int32']:
        # Use mean imputation for numerical type columns
        imputer = SimpleImputer(strategy='mean')
        train[col] = imputer.fit_transform(train[[col]]).ravel()
    else:
        print(f"Column '{col}' has unsupported data type: {train[col].dtype}")


# Find columns with 'object' dtype
object_cols = train.select_dtypes(include='object').columns

# Encode object columns using Label Encoding
from sklearn.preprocessing import LabelEncoder

for col in object_cols:
    le = LabelEncoder()
    train[col] = le.fit_transform(train[col])

train.dtypes

id                                         int64
Name                                       int64
Gender                                     int64
Age                                      float64
City                                       int64
Working Professional or Student            int64
Profession                                 int64
Academic Pressure                        float64
Work Pressure                            float64
CGPA                                     float64
Study Satisfaction                       float64
Job Satisfaction                         float64
Sleep Duration                             int64
Dietary Habits                             int64
Degree                                     int64
Have you ever had suicidal thoughts ?      int64
Work/Study Hours                         float64
Financial Stress                         float64
Family History of Mental Illness           int64
Depression                                 int64
dtype: object

In [ ]:
train.dtypes.unique()

array([dtype('int64'), dtype('float64')], dtype=object)

In [ ]:
correlations = train.corr()['Depression'].abs().sort_values(ascending=False)
print(correlations)

Depression                               1.000000
Age                                      0.564671
Working Professional or Student          0.520790
Have you ever had suicidal thoughts ?    0.349066
Profession                               0.283113
Academic Pressure                        0.270225
Financial Stress                         0.227232
Work/Study Hours                         0.191746
Dietary Habits                           0.143415
Work Pressure                            0.137823
Job Satisfaction                         0.107254
Study Satisfaction                       0.095577
Degree                                   0.045547
Name                                     0.028897
City                                     0.020915
Family History of Mental Illness         0.016502
Sleep Duration                           0.014454
CGPA                                     0.012361
Gender                                   0.008144
id                                       0.003944


In [ ]:
print(train['Depression'].dtype)

int64


In [ ]:
from sklearn.preprocessing import OrdinalEncoder

# Create OrdinalEncoder object
enc = OrdinalEncoder(categories=[['0', '1']])
# Fit and transform the data
train['Depression'] = enc.fit_transform(train[['Depression']]).astype(int)

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split

# Assuming 'train' is your DataFrame and 'sii' is your target variable
X = train.drop('Depression', axis=1)
y = train['Depression']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create and train the Decision Tree Classifier
tree_model = DecisionTreeClassifier(random_state=42)
tree_model.fit(X_train, y_train)

# Get feature importances
feature_importances = tree_model.feature_importances_


In [ ]:
from sklearn.metrics import accuracy_score, classification_report

# Make predictions on the test set
y_pred = tree_model.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy}")

# Print classification report (includes precision, recall, F1-score)
print(classification_report(y_test, y_pred))

Accuracy: 0.9024164889836531
              precision    recall  f1-score   support

           0       0.94      0.94      0.94     22986
           1       0.73      0.74      0.74      5154

    accuracy                           0.90     28140
   macro avg       0.84      0.84      0.84     28140
weighted avg       0.90      0.90      0.90     28140



In [ ]:
# prompt: random forest

# Create and train the Random Forest Classifier
rf_model = RandomForestClassifier(random_state=42)
rf_model.fit(X_train, y_train)

# Make predictions on the test set
y_pred_rf = rf_model.predict(X_test)

# Calculate accuracy
accuracy_rf = accuracy_score(y_test, y_pred_rf)
print(f"Random Forest Accuracy: {accuracy_rf}")

# Print classification report (includes precision, recall, F1-score)
print(classification_report(y_test, y_pred_rf))

Random Forest Accuracy: 0.9360341151385928
              precision    recall  f1-score   support

           0       0.95      0.97      0.96     22986
           1       0.85      0.79      0.82      5154

    accuracy                           0.94     28140
   macro avg       0.90      0.88      0.89     28140
weighted avg       0.93      0.94      0.94     28140



In [ ]:
# prompt: generate a code for predicting the depression 1 or 0 using LGBM ,XGBoost,CatBoost in ensemble method

!pip install lightgbm xgboost catboost

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report
from lightgbm import LGBMClassifier
from xgboost import XGBClassifier
from catboost import CatBoostClassifier
from sklearn.preprocessing import OrdinalEncoder # Import OrdinalEncoder for categorical feature encoding

# Assuming 'train' DataFrame is already preprocessed as in the previous code

# Separate features (X) and target variable (y)
X = train.drop('Depression', axis=1)
y = train['Depression']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize models
lgbm_model = LGBMClassifier(random_state=42)
xgb_model = XGBClassifier(random_state=42)
catboost_model = CatBoostClassifier(random_state=42, verbose=0) # Set verbose=0 to suppress training output

# Train models
lgbm_model.fit(X_train, y_train)
xgb_model.fit(X_train, y_train)
catboost_model.fit(X_train, y_train)

# ---->  Encode categorical features in the 'test' DataFrame
categorical_features = ['Name', 'Gender', 'City', 'Working Professional or Student', 'Profession', 'Sleep Duration', 'Dietary Habits', 'Degree', 'Have you ever had suicidal thoughts ?', 'Family History of Mental Illness']
enc = OrdinalEncoder() # Create an OrdinalEncoder object
test[categorical_features] = enc.fit_transform(test[categorical_features]) # Fit and transform the categorical features in 'test'


lgbm_pred = lgbm_model.predict(X_test)
xgb_pred = xgb_model.predict(X_test)
catboost_pred = catboost_model.predict(X_test)

# Ensemble predictions (simple averaging)
ensemble_pred = np.round((lgbm_pred + xgb_pred + catboost_pred) / 3).astype(int)

# Evaluate ensemble
accuracy_ensemble = accuracy_score(y_test, ensemble_pred)
print(f"Ensemble Accuracy: {accuracy_ensemble}")
print(classification_report(y_test, ensemble_pred))

[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Number of positive: 20413, number of negative: 92147
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.011603 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 919
[LightGBM] [Info] Number of data points in the train set: 112560, number of used features: 19
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.181352 -> initscore=-1.507213
[LightGBM] [Info] Start training from score -1.507213
Ensemble Accuracy: 0.9393390191897655
              precision    recall  f1-score   support

           0       0.96      0.97      0.96     22986
           1       0.84      0.82      0.83      5154

    accuracy                           0.94     28140
   macro avg       0.90      0.89      0.90     28140
weighted avg       0.94      0.94      0.94     2814

In [ ]:
# prompt: count number of elements in ensemble_pred

print(len(ensemble_pred))

28140


In [ ]:
# Assuming 'test' is your DataFrame

# Import necessary libraries
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.compose import ColumnTransformer

# Identify categorical features for encoding
categorical_features = ['Name', 'Gender', 'City', 'Working Professional or Student',
                         'Profession', 'Sleep Duration', 'Dietary Habits', 'Degree',
                         'Have you ever had suicidal thoughts ?', 'Family History of Mental Illness']

# Create a ColumnTransformer to apply different encodings
# Passthrough all columns in X

In [ ]:
lgbm_pred = lgbm_model.predict(test)
xgb_pred = xgb_model.predict(test)
catboost_pred = catboost_model.predict(test)

# Ensemble predictions (simple averaging)
ensemble_pred = np.round((lgbm_pred + xgb_pred + catboost_pred) / 3).astype(int)


In [ ]:
import lightgbm as lgb
from xgboost import XGBClassifier  # or XGBRegressor for regression
from catboost import CatBoostClassifier  # or CatBoostRegressor for regression
from sklearn.model_selection import KFold
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler

def ensemble_model(X_train, y_train, X_test):

    # Define base models
    lgbm_model = lgb.LGBMClassifier(objective='binary')  # adjust objective based on task
    xgb_model = XGBClassifier()
    cat_model = CatBoostClassifier()

    # Create KFold for cross-validation
    kf = KFold(n_splits=5, shuffle=True, random_state=42)

    # Initialize list to store predictions from each model
    lgbm_preds = []
    xgb_preds = []
    cat_preds = []

    # Train each model with KFold and store predictions
    for train_index, val_index in kf.split(X_train):
        X_tr, X_val = X_train.iloc[train_index], X_train.iloc[val_index]
        y_tr, y_val = y_train.iloc[train_index], y_train.iloc[val_index]

        # Fit and predict for LightGBM
        lgbm_model.fit(X_tr, y_tr)
        lgbm_preds.append(lgbm_model.predict_proba(X_val)[:, 1])

        # Fit and predict for XGBoost
        xgb_model.fit(X_tr, y_tr)
        xgb_preds.append(xgb_model.predict_proba(X_val)[:, 1])

        # Fit and predict for CatBoost
        cat_model.fit(X_tr, y_tr)
        cat_preds.append(cat_model.predict_proba(X_val)[:, 1])

    # Average predictions from each model
    final_preds = (np.array(lgbm_preds) + np.array(xgb_preds) + np.array(cat_preds)) / 3

    # Predict on test data
    lgbm_test_pred = lgbm_model.predict_proba(X_test)[:, 1]
    xgb_test_pred = xgb_model.predict_proba(X_test)[:, 1]
    cat_test_pred = cat_model.predict_proba(X_test)[:, 1]
    ensemble_test_pred = (lgbm_test_pred + xgb_test_pred + cat_test_pred) / 3

    return final_preds, ensemble_test_pred

In [ ]:
ensemble_pred

array([0, 0, 0, ..., 0, 1, 0])

In [ ]:
# prompt: create a sample_submission.csv for ensemle_pred predictions with two a columns id and depression.fix an error:---------------------------------------------------------------------------
# ValueError                                Traceback (most recent call last)
# Cell In[63], line 13
#       7 import numpy as np
#       9 # Assuming 'ensemble_pred' is your NumPy array of predictions
#      10 # and 'test' DataFrame contains the 'id' column
#      11
#      12 # Create a sample submission DataFrame
# ---> 13 submission = pd.DataFrame({'id': test['id'], 'Depression': ensemble_pred})
#      15 # Save the DataFrame to a CSV file
#      16 submission.to_csv('sample_submission.csv', index=False)
# File /opt/conda/lib/python3.10/site-packages/pandas/core/frame.py:778, in DataFrame.__init__(self, data, index, columns, dtype, copy)
#     772     mgr = self._init_mgr(
#     773         data, axes={"index": index, "columns": columns}, dtype=dtype, copy=copy
#     774     )
#     776 elif isinstance(data, dict):
#     777

import pandas as pd
import numpy as np

# Assuming 'ensemble_pred' is your NumPy array of predictions
# and 'test' DataFrame contains the 'id' column

# Create a sample submission DataFrame
submission = pd.DataFrame({'id': test['id'], 'Depression': ensemble_pred})

# Ensure 'Depression' column is of integer type
submission['Depression'] = submission['Depression'].astype(int)

# Save the DataFrame to a CSV file
submission.to_csv('sample_submission.csv', index=False)

In [ ]:
# prompt: Use the model to get a prediction of a child

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OrdinalEncoder
from lightgbm import LGBMClassifier
from xgboost import XGBClassifier
from catboost import CatBoostClassifier
from sklearn.metrics import accuracy_score, classification_report

# Load the preprocessed training data (assuming it's named 'train_preprocessed.csv')
try:
    train = pd.read_csv('/content/train_preprocessed.csv')
except FileNotFoundError:
    print("Error: 'train_preprocessed.csv' not found. Please ensure the preprocessed training data is saved with this name.")
    exit()

# Assuming 'train' is your DataFrame and 'sii' is your target variable
X = train.drop('Depression', axis=1)
y = train['Depression']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize models (you might need to adjust hyperparameters)
lgbm_model = LGBMClassifier(random_state=42)
xgb_model = XGBClassifier(random_state=42)
catboost_model = CatBoostClassifier(random_state=42, verbose=0)

# Train models
lgbm_model.fit(X_train, y_train)
xgb_model.fit(X_train, y_train)
catboost_model.fit(X_train, y_train)

# Load the test data
try:
    test = pd.read_csv('/content/test (2).csv')
except FileNotFoundError:
    print("Error: 'test (2).csv' not found. Please ensure the test data is in the correct location.")
    exit()

# Preprocess the test data (must mirror the preprocessing done for the training data)
# ... (your preprocessing code here) ...

# Example of simple preprocessing: impute missing values and label encode
for col in test.columns:
    if test[col].isnull().any():
        if test[col].dtype == 'object':
            test[col].fillna(test[col].mode()[0], inplace=True)
        else:
            test[col].fillna(test[col].mean(), inplace=True)

object_cols = test.select_dtypes(include='object').columns

for col in object_cols:
    le = LabelEncoder()
    test[col] = le.fit_transform(test[col])


# Make predictions on the test set using the ensemble
lgbm_pred_test = lgbm_model.predict(test)
xgb_pred_test = xgb_model.predict(test)
catboost_pred_test = catboost_model.predict(test)

# Ensemble predictions (simple averaging)
ensemble_pred_test = np.round((lgbm_pred_test + xgb_pred_test + catboost_pred_test) / 3).astype(int)

# Print predictions for first few rows (or as needed)
print("Predictions for the first 5 children in the test set:")
print(ensemble_pred_test[:5])

Error: 'train_preprocessed.csv' not found. Please ensure the preprocessed training data is saved with this name.
[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Number of positive: 20413, number of negative: 92147
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.019248 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 919
[LightGBM] [Info] Number of data points in the train set: 112560, number of used features: 19
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.181352 -> initscore=-1.507213
[LightGBM] [Info] Start training from score -1.507213


<ipython-input-26-a2f44083c296>:51: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  test[col].fillna(test[col].mode()[0], inplace=True)
<ipython-input-26-a2f44083c296>:53: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try

Predictions for the first 5 children in the test set:
[0 0 0 1 0]


The above code is ensmeble method of LGBM ,XG Boost and Cat Boost to get the efficient accuracy.The accuracy i got is 94%
